File: src/sdks/typescript/src/commands.ts
import createClient, { type Client } from 'openapi-fetch'

import { unwrap } from './api'
import type {
  components as dataplane,
  paths as dataplanePaths,
} from './dataplane/schema.gen'
import {
  capabilityToken,
  dataPlaneOrigin,
  DEFAULT_AGENT_PORT,
  openDataPlaneWebSocket,
  withFreshToken,
  type DataPlaneOptions,
  type TokenSource,
  type WebSocketFactory,
} from './dataplane'
import { defaultHeaders } from './api/metadata'
import {
  createHttpFetch,
  RequestPipeline,
  type CallInit,
  type CallSpec,
  type RequestOptions,
} from './http'
import {
  isNonNegativeInteger,
  isPositiveInteger,
  parseJsonObject,
  StreamingEvents,
  StreamingProtocolError,
  type StreamOptions,
} from './streaming'

/** A finished command: exit code and captured output. */
export type CommandResult = dataplane['schemas']['CommandResult']

/** One process the agent supervises inside the sandbox. */
export type Process = dataplane['schemas']['Process']

/** A snapshot of retained output for a background process. */
export type ProcessLogs = dataplane['schemas']['ProcessLogs']

/** Terminal dimensions used when starting or resizing a PTY process. */
export type PtySize = dataplane['schemas']['PtySize']

/** Signals {@link Commands.kill} can deliver. */
export type Signal = 'SIGTERM' | 'SIGKILL' | 'SIGINT' | 'SIGHUP'

/** Options for running a command. */
export interface RunCommandOptions extends RequestOptions {
  /** Arguments passed to the program. */
  args?: string[]
  /**
   * Working directory for the command.
   *
   * @default the default user's home directory
   */
  cwd?: string
  /** Environment variables set for this command only. */
  env?: Record<string, string>
  /** Bytes written to the command's stdin before it is closed. */
  stdin?: string
  /**
   * Kill the command if it has not exited after this many seconds.
   *
   * @default the agent's default
   */
  timeoutSeconds?: number
}

/** Options for starting a supervised background process. */
export interface StartProcessOptions extends RequestOptions {
  /** Arguments passed to the program. */
  args?: string[]
  /** Working directory for the process. */
  cwd?: string
  /** Environment variables set for this process only. */
  env?: Record<string, string>
  /** Allocate a PTY with these initial dimensions. */
  pty?: PtySize
}

/**
 * Options for attaching to retained and live process output.
 *
 * `requestTimeoutMs` bounds the WebSocket handshake, not the lifetime
 * of the stream; `signal` closes the stream.
 */
export interface AttachProcessOptions extends RequestOptions {
  /** Last output sequence already observed. */
  after?: number
}

/** A validated event received from an attached process stream. */
export type ProcessStreamEvent =
  | { type: 'stdout'; sequence: number; data: string }
  | { type: 'stderr'; sequence: number; data: string }
  | { type: 'exit'; sequence: number; exitCode: number }
  | { type: 'gap'; fromSequence: number; toSequence: number }

function parseProcessStreamEvent(
  data: unknown,
): ProcessStreamEvent | undefined {
  const message = parseJsonObject(data)
  if (typeof message['type'] !== 'string') {
    throw new StreamingProtocolError(
      'process stream frame type must be a string',
    )
  }
  switch (message['type']) {
    case 'stdout':
    case 'stderr':
      if (
        !isPositiveInteger(message['sequence']) ||
        typeof message['data'] !== 'string'
      ) {
        throw new StreamingProtocolError(
          `malformed known ${message['type']} frame`,
        )
      }
      return {
        type: message['type'],
        sequence: message['sequence'],
        data: message['data'],
      }
    case 'exit':
      if (
        !isPositiveInteger(message['sequence']) ||
        typeof message['exit_code'] !== 'number' ||
        !Number.isInteger(message['exit_code'])
      ) {
        throw new StreamingProtocolError('malformed known exit frame')
      }
      return {
        type: 'exit',
        sequence: message['sequence'],
        exitCode: message['exit_code'],
      }
    case 'gap':
      if (
        !isPositiveInteger(message['from_sequence']) ||
        !isPositiveInteger(message['to_sequence'])
      ) {
        throw new StreamingProtocolError('malformed known gap frame')
      }
      return {
        type: 'gap',
        fromSequence: message['from_sequence'],
        toSequence: message['to_sequence'],
      }
    default:
      return undefined
  }
}

/** A live process WebSocket with typed events and interactive controls. */
export class ProcessAttachment extends StreamingEvents<ProcessStreamEvent> {
  protected parse(data: unknown): ProcessStreamEvent | undefined {
    return parseProcessStreamEvent(data)
  }

  /** Writes text to the process stdin. */
  sendInput(data: string): void {
    this.send({ type: 'input', data })
  }

  /** Closes the process stdin without closing the output stream. */
  closeStdin(): void {
    this.send({ type: 'close_stdin' })
  }

  /** Changes the dimensions of an attached PTY. */
  resize(cols: number, rows: number): void {
    this.send({ type: 'resize', cols, rows })
  }

  /** Delivers a signal to the process through the attached stream. */
  signal(signal: Signal): void {
    this.send({ type: 'signal', signal })
  }
}

/** Ergonomic operations bound to one supervised process. */
export class ProcessHandle {
  constructor(
    private readonly commands: Commands,
    public info: Process,
  ) {}

  /** The process identifier. */
  get pid(): number {
    return this.info.pid
  }

  /** Refreshes and returns process metadata. */
  async refresh(options?: RequestOptions): Promise<Process> {
    this.info = await this.commands.getProcess(this.pid, options)
    return this.info
  }

  /** Reads the retained output snapshot. */
  logs(after?: number, options?: RequestOptions): Promise<ProcessLogs> {
    return this.commands.logs(this.pid, after, options)
  }

  /** Attaches to retained and live output. */
  attach(options?: AttachProcessOptions): ProcessAttachment {
    return this.commands.attach(this.pid, options)
  }

  /** Sends a signal to this process. */
  kill(signal?: Signal, options?: RequestOptions): Promise<void> {
    return this.commands.kill(this.pid, signal, options)
  }
}

/**
 * Module for running commands inside the sandbox, available as
 * `sandbox.commands`.
 *
 * @example
 * ```ts
 * const sandbox = await client.sandboxes.create({ template: 'base' })
 * const result = await sandbox.commands.run('echo', { args: ['hello'] })
 * console.log(result.stdout) // hello
 * ```
 */
export class Commands {
  private client?: Client<dataplanePaths>
  private clientToken?: string
  private readonly agentPort: number
  private readonly fetch: typeof globalThis.fetch
  private readonly webSocket?: WebSocketFactory
  private readonly requests: RequestPipeline

  /**
   * Creates a command module bound to a sandbox handle.
   *
   * @param sandbox Source of the current sandbox identity and token.
   * @param options Data-plane connection settings and request defaults.
   */
  constructor(
    private readonly sandbox: TokenSource,
    options?: DataPlaneOptions,
  ) {
    this.agentPort = options?.agentPort ?? DEFAULT_AGENT_PORT
    this.requests = new RequestPipeline(options)
    this.fetch = createHttpFetch(this.requests.requestTimeoutMs, options?.fetch)
    this.webSocket = options?.webSocket
  }

  /** Returns a client for the current token, rebuilding it after token rotation. */
  private dataplane(): Client<dataplanePaths> {
    const token = capabilityToken(this.sandbox)
    if (this.client === undefined || this.clientToken !== token) {
      this.client = createClient<dataplanePaths>({
        baseUrl: dataPlaneOrigin(this.sandbox, this.agentPort),
        headers: { ...defaultHeaders, 'X-Sandbox-Token': token },
        fetch: this.fetch,
      })
      this.clientToken = token
    }
    return this.client
  }

  /**
   * Runs one data-plane call with retry and automatic recovery from a
   * rejected capability token.
   */
  private call<T>(
    spec: CallSpec,
    run: (client: Client<dataplanePaths>, init: CallInit) => Promise<T>,
  ): Promise<T> {
    return withFreshToken(this.sandbox, this.requests.logger, () => {
      // Resolved outside the retry loop: an unarmed handle is a caller
      // mistake, not a failure worth replaying.
      const client = this.dataplane()
      return this.requests.send(spec, (init) => run(client, init))
    })
  }

  /**
   * Run a command to completion and return its captured output.
   *
   * Blocks until the command exits or `timeoutSeconds` elapses. The
   * response marks output that exceeded the capture limit as truncated.
   *
   * @param command The program to run.
   * @param opts Options for running the command.
   * @returns The command result.
   * @throws {CommandTimeoutError} When the command exceeds `timeoutSeconds`.
   *
   * @example
   * ```ts
   * const result = await sandbox.commands.run('python3', {
   *   args: ['-c', 'print(6 * 7)'],
   *   timeoutSeconds: 60,
   * })
   * console.log(result.stdout) // 42
   * ```
   */
  async run(command: string, opts?: RunCommandOptions): Promise<CommandResult> {
    return this.call(
      { operation: 'POST /commands', idempotent: false, options: opts },
      async (client, init) =>
        unwrap(
          await client.POST('/commands', {
            body: {
              command,
              ...(opts?.args !== undefined && { args: opts.args }),
              ...(opts?.cwd !== undefined && { cwd: opts.cwd }),
              ...(opts?.env !== undefined && { env: opts.env }),
              ...(opts?.stdin !== undefined && { stdin: opts.stdin }),
              ...(opts?.timeoutSeconds !== undefined && {
                timeout_seconds: opts.timeoutSeconds,
              }),
            },
            ...init,
          }),
        ),
    )
  }

  /**
   * Starts a supervised process and returns a handle immediately.
   *
   * Pass `pty` to allocate a terminal; then call `attach()` on the
   * returned handle to interact with it.
   */
  async start(
    command: string,
    opts?: StartProcessOptions,
  ): Promise<ProcessHandle> {
    const process = await this.call(
      { operation: 'POST /processes', idempotent: false, options: opts },
      async (client, init) =>
        unwrap(
          await client.POST('/processes', {
            body: {
              command,
              ...(opts?.args !== undefined && { args: opts.args }),
              ...(opts?.cwd !== undefined && { cwd: opts.cwd }),
              ...(opts?.env !== undefined && { env: opts.env }),
              ...(opts?.pty !== undefined && { pty: opts.pty }),
            },
            ...init,
          }),
        ),
    )
    return new ProcessHandle(this, process)
  }

  /**
   * Gets one supervised process.
   *
   * @param pid Process ID from {@link Commands.listProcesses}.
   * @param options Per-call timeout, signal, and headers.
   */
  async get(pid: number, options?: RequestOptions): Promise<ProcessHandle> {
    return new ProcessHandle(this, await this.getProcess(pid, options))
  }

  /**
   * Gets raw metadata for one supervised process.
   *
   * @param pid Process ID from {@link Commands.listProcesses}.
   * @param options Per-call timeout, signal, and headers.
   */
  async getProcess(pid: number, options?: RequestOptions): Promise<Process> {
    return this.call(
      { operation: 'GET /processes/{pid}', idempotent: true, options },
      async (client, init) =>
        unwrap(
          await client.GET('/processes/{pid}', {
            params: { path: { pid } },
            ...init,
          }),
        ),
    )
  }

  /**
   * Reads retained output for a supervised process.
   *
   * @param pid Process ID from {@link Commands.listProcesses}.
   * @param after Last output sequence already observed.
   * @param options Per-call timeout, signal, and headers.
   */
  async logs(
    pid: number,
    after?: number,
    options?: RequestOptions,
  ): Promise<ProcessLogs> {
    return this.call(
      { operation: 'GET /processes/{pid}/logs', idempotent: true, options },
      async (client, init) =>
        unwrap(
          await client.GET('/processes/{pid}/logs', {
            params: {
              path: { pid },
              query: after !== undefined ? { after } : {},
            },
            ...init,
          }),
        ),
    )
  }

  /**
   * Attaches to retained and live process output over WebSocket.
   *
   * The handle must already hold a capability token: unlike the HTTP
   * calls in this module, a handshake the agent rejects cannot be
   * distinguished from any other socket failure, so the SDK cannot
   * re-mint and reconnect on its own.
   *
   * @param pid Process ID from {@link Commands.listProcesses}.
   * @param options Sequence to resume from, handshake timeout, and signal.
   */
  attach(pid: number, options?: AttachProcessOptions): ProcessAttachment {
    const after = options?.after ?? 0
    if (!isNonNegativeInteger(after)) {
      throw new RangeError('after must be a non-negative integer')
    }
    const socket = openDataPlaneWebSocket(
      this.sandbox,
      this.agentPort,
      this.webSocket,
      `/processes/${pid}/stream`,
      { after },
    )
    return new ProcessAttachment(socket, this.streamOptions(options))
  }

  /**
   * List the processes the agent supervises inside the sandbox.
   *
   * @param options Per-call timeout, signal, and headers.
   * @returns The supervised processes.
   */
  async listProcesses(options?: RequestOptions): Promise<Process[]> {
    const page = await this.call(
      { operation: 'GET /processes', idempotent: true, options },
      async (client, init) =>
        unwrap(await client.GET('/processes', { ...init })),
    )
    return page.items
  }

  /**
   * Send a signal to a process.
   *
   * @param pid Process ID from {@link Commands.listProcesses}.
   * @param signal Signal to deliver. Defaults to `SIGTERM`.
   * @param options Per-call timeout, signal, and headers.
   */
  async kill(
    pid: number,
    signal?: Signal,
    options?: RequestOptions,
  ): Promise<void> {
    await this.call(
      { operation: 'DELETE /processes/{pid}', idempotent: true, options },
      async (client, init) =>
        unwrap(
          await client.DELETE('/processes/{pid}', {
            params: {
              path: { pid },
              query: signal !== undefined ? { signal } : {},
            },
            ...init,
          }),
        ),
    )
  }

  /** Maps per-call overrides onto a stream's connection settings. */
  private streamOptions(options?: RequestOptions): StreamOptions {
    return {
      handshakeTimeoutMs:
        options?.requestTimeoutMs ?? this.requests.requestTimeoutMs,
      ...(options?.signal !== undefined && { signal: options.signal }),
    }
  }
}

File: src/sdks/rust/src/commands.rs
//! The sandbox data plane: run-to-completion commands and process
//! management, from `src/contracts/dataplane.yaml`.
//!
//! Served at the sandbox's own hostname, off the control plane:
//! available whenever the sandbox is running, even when the control
//! plane is not.

use std::pin::Pin;
use std::task::{Context, Poll};

use futures_util::Stream;
use reqwest::Method;
use serde::Serialize;

use crate::client::AgentTarget;
use crate::dataplane::DataPlane;
use crate::error::Error;
use crate::http::Http;
use crate::models;
use crate::models::{
    CommandResult, Process, ProcessList, ProcessLogs, ProcessStreamEvent, RunCommandRequest,
    StartProcessRequest,
};
use crate::streaming::EventSocket;
use crate::token::TokenSource;

/// A signal deliverable through [`Commands::kill`].
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum Signal {
    /// Graceful termination — the server default when none is sent.
    Term,
    /// Immediate, uncatchable kill.
    Kill,
    /// Interrupt, as from Ctrl-C.
    Int,
    /// Hangup.
    Hup,
}

impl Signal {
    /// The wire name, as used in the `signal` query parameter.
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Term => "SIGTERM",
            Self::Kill => "SIGKILL",
            Self::Int => "SIGINT",
            Self::Hup => "SIGHUP",
        }
    }
}

/// Command execution inside one sandbox. Obtained from
/// [`crate::Sandbox::commands`].
///
/// Every request carries the sandbox's capability token as
/// `X-Sandbox-Token`. The token is bound to the sandbox's epoch, so a
/// `Commands` value outlives its token whenever the sandbox is resumed;
/// a rejected token is re-minted through the control plane once and the
/// call is replayed, so the value keeps working across a resume.
///
/// # Examples
///
/// ```no_run
/// use fissionplane::models::RunCommandRequest;
///
/// # async fn demo(sandbox: fissionplane::Sandbox) -> Result<(), fissionplane::Error> {
/// let result = sandbox
///     .commands()?
///     .run(RunCommandRequest {
///         command: "python".to_owned(),
///         args: Some(vec!["-V".to_owned()]),
///         ..Default::default()
///     })
///     .await?;
/// println!("{}", result.stdout);
/// # Ok(())
/// # }
/// ```
#[derive(Clone, Debug)]
pub struct Commands {
    data: DataPlane,
}

impl Commands {
    pub(crate) fn new(
        http: &Http,
        agent: &AgentTarget,
        info: &models::Sandbox,
        credential: TokenSource,
    ) -> Result<Self, Error> {
        Ok(Self {
            data: DataPlane::new(http, agent, info, credential)?,
        })
    }

    /// Start the command inside the sandbox and block until it exits
    /// or its timeout elapses. Output is captured and returned in one
    /// document, truncated at the advertised limit.
    ///
    /// A command that overruns `timeout_seconds` is killed and the
    /// call fails with [`Error::Api`] carrying status 408.
    ///
    /// # Errors
    ///
    /// Returns [`Error::Api`] for a rejected command, including a
    /// timeout, or [`Error::Http`] if transport or decoding fails.
    pub async fn run(&self, request: RunCommandRequest) -> Result<CommandResult, Error> {
        let builder = self.data.request(Method::POST, "/commands").json(&request);
        self.data.send_json(builder).await
    }

    /// Starts a supervised background process and returns a bound handle.
    ///
    /// Set [`StartProcessRequest::pty`] to allocate a pseudo-terminal
    /// before attaching.
    ///
    /// # Errors
    ///
    /// Returns [`Error::Api`] if the process cannot be started or
    /// [`Error::Http`] if transport or decoding fails.
    pub async fn start(&self, request: StartProcessRequest) -> Result<ProcessHandle, Error> {
        let builder = self.data.request(Method::POST, "/processes").json(&request);
        let process = self.data.send_json(builder).await?;
        Ok(ProcessHandle::new(self.clone(), process))
    }

    /// Gets a supervised process and returns a bound handle.
    ///
    /// # Errors
    ///
    /// Returns [`Error::Api`] when the process is unavailable or
    /// [`Error::Http`] if transport or decoding fails.
    pub async fn get(&self, pid: u32) -> Result<ProcessHandle, Error> {
        Ok(ProcessHandle::new(
            self.clone(),
            self.get_process(pid).await?,
        ))
    }

    /// Gets raw metadata for one supervised process.
    ///
    /// # Errors
    ///
    /// Returns [`Error::Api`] when the process is unavailable or
    /// [`Error::Http`] if transport or decoding fails.
    pub async fn get_process(&self, pid: u32) -> Result<Process, Error> {
        let request = self.data.request(Method::GET, &format!("/processes/{pid}"));
        self.data.send_json(request).await
    }

    /// Reads a snapshot of retained process output after `after`.
    ///
    /// # Errors
    ///
    /// Returns [`Error::Api`] when the process is unavailable or
    /// [`Error::Http`] if transport or decoding fails.
    pub async fn logs(&self, pid: u32, after: Option<u64>) -> Result<ProcessLogs, Error> {
        let mut request = self
            .data
            .request(Method::GET, &format!("/processes/{pid}/logs"));
        if let Some(after) = after {
            request = request.query(&[("after", after)]);
        }
        self.data.send_json(request).await
    }

    /// Attaches to retained and live process output.
    ///
    /// `after` is the last output sequence already observed.
    ///
    /// A handshake the agent rejects for the token is retried once with
    /// a freshly minted one, so attaching works across a resume.
    ///
    /// # Errors
    ///
    /// Returns [`Error::WebSocket`] if the handshake fails,
    /// [`Error::Protocol`] if the server selects the wrong subprotocol,
    /// or [`Error::WaitTimeout`] if the handshake outlives the client's
    /// request timeout.
    pub async fn attach(&self, pid: u32, after: u64) -> Result<ProcessAttachment, Error> {
        let url = self
            .data
            .base()
            .join(&format!("/processes/{pid}/stream"))
            .map_err(|source| Error::Config(format!("process stream URL: {source}")))?;
        let socket = self
            .data
            .connect(url, &[("after", after.to_string())])
            .await?;
        Ok(ProcessAttachment { socket })
    }

    /// The processes the agent supervises inside the sandbox.
    ///
    /// Returns the current process list.
    ///
    /// # Errors
    ///
    /// Returns [`Error::Api`] if the agent rejects the request or
    /// [`Error::Http`] if transport or decoding fails.
    pub async fn list_processes(&self) -> Result<Vec<Process>, Error> {
        let request = self.data.request(Method::GET, "/processes");
        let list: ProcessList = self.data.send_json(request).await?;
        Ok(list.items)
    }

    /// Send `signal` to the process. `None` means the server default,
    /// SIGTERM.
    ///
    /// # Errors
    ///
    /// Returns [`Error::Api`] if the process cannot be signalled or
    /// [`Error::Http`] if transport fails.
    pub async fn kill(&self, pid: u32, signal: Option<Signal>) -> Result<(), Error> {
        let mut request = self
            .data
            .request(Method::DELETE, &format!("/processes/{pid}"));
        if let Some(signal) = signal {
            request = request.query(&[("signal", signal.as_str())]);
        }
        self.data.send_no_content(request).await
    }
}

/// Ergonomic operations bound to one supervised process.
#[derive(Clone, Debug)]
pub struct ProcessHandle {
    commands: Commands,
    /// Latest process metadata fetched by this handle.
    pub info: Process,
}

impl ProcessHandle {
    fn new(commands: Commands, info: Process) -> Self {
        Self { commands, info }
    }

    /// Returns the process identifier.
    pub fn pid(&self) -> u32 {
        self.info.pid
    }

    /// Refreshes and returns process metadata.
    ///
    /// # Errors
    ///
    /// Returns the same errors as [`Commands::get_process`].
    pub async fn refresh(&mut self) -> Result<&Process, Error> {
        self.info = self.commands.get_process(self.info.pid).await?;
        Ok(&self.info)
    }

    /// Reads retained output after `after`.
    ///
    /// # Errors
    ///
    /// Returns the same errors as [`Commands::logs`].
    pub async fn logs(&self, after: Option<u64>) -> Result<ProcessLogs, Error> {
        self.commands.logs(self.info.pid, after).await
    }

    /// Attaches to retained and live output after `after`.
    ///
    /// # Errors
    ///
    /// Returns the same errors as [`Commands::attach`].
    pub async fn attach(&self, after: u64) -> Result<ProcessAttachment, Error> {
        self.commands.attach(self.info.pid, after).await
    }

    /// Sends a signal to the process.
    ///
    /// # Errors
    ///
    /// Returns the same errors as [`Commands::kill`].
    pub async fn kill(&self, signal: Option<Signal>) -> Result<(), Error> {
        self.commands.kill(self.info.pid, signal).await
    }
}

/// An attached process WebSocket with typed events and interactive controls.
#[derive(Debug)]
pub struct ProcessAttachment {
    socket: EventSocket,
}

impl ProcessAttachment {
    /// Writes UTF-8 text to process stdin.
    ///
    /// # Errors
    ///
    /// Returns [`Error::WebSocket`] when sending fails.
    pub async fn send_input(&mut self, data: impl Into<String>) -> Result<(), Error> {
        self.socket
            .send(&ProcessClientMessage::Input { data: data.into() })
            .await
    }

    /// Closes process stdin without closing the output stream.
    ///
    /// # Errors
    ///
    /// Returns [`Error::WebSocket`] when sending fails.
    pub async fn close_stdin(&mut self) -> Result<(), Error> {
        self.socket.send(&ProcessClientMessage::CloseStdin).await
    }

    /// Changes the dimensions of an attached pseudo-terminal.
    ///
    /// # Errors
    ///
    /// Returns [`Error::WebSocket`] when sending fails.
    pub async fn resize(&mut self, cols: u16, rows: u16) -> Result<(), Error> {
        self.socket
            .send(&ProcessClientMessage::Resize { cols, rows })
            .await
    }

    /// Delivers a signal through the attached stream.
    ///
    /// # Errors
    ///
    /// Returns [`Error::WebSocket`] when sending fails.
    pub async fn signal(&mut self, signal: Signal) -> Result<(), Error> {
        self.socket
            .send(&ProcessClientMessage::Signal {
                signal: signal.as_str(),
            })
            .await
    }

    /// Closes the WebSocket.
    ///
    /// # Errors
    ///
    /// Returns [`Error::WebSocket`] when closing fails.
    pub async fn close(&mut self) -> Result<(), Error> {
        self.socket.close().await
    }
}

impl Stream for ProcessAttachment {
    type Item = Result<ProcessStreamEvent, Error>;

    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        self.get_mut().socket.poll_event(
            cx,
            &["stdout", "stderr", "exit", "gap"],
            "process stream",
            valid_process_event,
        )
    }
}

fn valid_process_event(event: &ProcessStreamEvent) -> bool {
    match event {
        ProcessStreamEvent::Stdout { sequence, .. }
        | ProcessStreamEvent::Stderr { sequence, .. }
        | ProcessStreamEvent::Exit { sequence, .. } => *sequence > 0,
        ProcessStreamEvent::Gap {
            from_sequence,
            to_sequence,
        } => *from_sequence > 0 && *to_sequence > 0,
    }
}

#[derive(Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
enum ProcessClientMessage<'a> {
    Input { data: String },
    CloseStdin,
    Resize { cols: u16, rows: u16 },
    Signal { signal: &'a str },
}
